### ISSUE TITLE
Possible race-condition in process stash

### ISSUE THREAD

**Issue by MartinFlores751** [irods/irods#7297]
- [ ] main
- [ ] 4-3-stable

---

## Bug Report

https://github.com/irods/irods/blob/21253de25be97fdc5ca6c458fab0759c0ee9cfc2/lib/core/src/process_stash.cpp#L41-L51

Currently, process stash has the possibility to have a race condition, as the `boost::any*` from `process_stash::find(...)` is unprotected.

```C++
auto is_state_valid{[](const std::string& _in_state) {
    // Search to see if state exists
    const auto* expire_time{boost::any_cast<std::chrono::steady_clock::time_point>(irods::process_stash::find(_in_state))};

    // It exists!!! Let's see what we can do...
    if (expire_time != nullptr) {
        // Ensure state is not expired...
        const auto now{std::chrono::steady_clock::now()};
        const auto is_not_expired{now < *expire_time};

        // Remove item from valid states
        irods::process_stash::erase(_in_state);

        // Return state validity
        return is_not_expired;
    }

    // Does not exist...
    return false;
}};
```
In the above example from the HTTP API, assuming two or more threads are accessing the same item using `process_stash::find(...)`, it is possible for one thread to retrieve the pointer, and another to delete the data being pointed to using `process_stash::erase(...)`. This would then lead to undefined behavior, as a thread may have already retrieved a pointer, but the pointed to data has been deleted.

**Comment by MartinFlores751** [comment]
Current ideas:
- Return a copy of the data, instead of a pointer
- Use a `std::shared_ptr` in the process stash map
- Have a mechanism to mark that an element is being used

**Comment by trel** [comment]
`std::shared_ptr` seems... the easiest and doesn't make a copy?

**Comment by trel** [comment]
Hmm, that would just make the window a little smaller I guess.    Probably a lock is the right answer.

**Comment by korydraughn** [comment]
Maintaining a separate list/mapping keeps the process stash impl simple and pushes the decision on how the developer wants to manage who is using what.

**Comment by korydraughn** [comment]
I'm starting to think this is out-of-scope for the `process_stash`.

The situation described exists for the standard c++ containers as well.

With that said, I think the most elegant solution is to introduce a `pop`/`take`/`steal`/`extract`/`remove` function.

This function would acquire the lock, remove the element from the stash, and return it to the caller. This removes the possibility of a race condition entirely and avoids unnecessary complexity.
```cpp
auto irods::process_stash::remove(const std::string& key) -> std::optional<restorable_entry>;
```

If the caller wants to add the item back, they can call `insert`. Now the question becomes, if they return it for some reason, should it retain the original key?

**Comment by MartinFlores751** [comment]
> `std::shared_ptr` seems... the easiest and doesn't make a copy?

Looking back, I think a shared pointer would not be the way to go, as I don't think they're thread safe.

**Comment by MartinFlores751** [comment]
> So... just lock() before find()?
> 
> A global/singleton vector/map of semaphores... kept outside the process stash.
> 
> Should provide a place to store that a key is in use.
> 
> So the pattern becomes...
> 
> lock, find, use, erase, unlock

Not the biggest fan of that. It's taking care of a task that should be handled by process stash itself, encapsulation being the idea.

Process stash, for the most part, is handling the locking. It's just the pointer circumvents it.

**Comment by MartinFlores751** [comment]
> Current ideas:
> 
>     * Return a copy of the data, instead of a pointer
> 
>     * Use a `std::shared_ptr` in the process stash map
> 
>     * Have a mechanism to mark that an element is being used

I'm curious if the performance penalty for copying is that great, since the solution is to remove characters at most?

It might not be the most performant way to handle things, but it leaves virtually everything about process stash the same, and removes the race condition.

**Comment by MartinFlores751** [comment]
> I think introducing a `pop`/`take`/`steal`/`extract`/`remove` function would give you a way to solve your issue.
> 
> This function would acquire the lock, remove the element from the stash, and return it to the caller. This removes the possibility of a race condition entirely and avoids unnecessary complexity.
> 
> ```c++
> auto irods::process_stash::remove(const std::string& key) -> std::optional<restorable_entry>;
> ```
> 
> If the caller wants to add the item back, they can call `insert`. Now the question becomes, if they return it for some reason, should it retain the original key?

I like the idea of this. My concern is what are the common operations of process stash? If it's mostly 'read-only' operations, then we have to do 2x the work (read then put back where it belongs).

Not sure if this may be a common case for process stash (as it is currently), but if people want to access the same element, then it may not exist for one if we "pop" it.

This made me think, what if we have a `peek(...)` operation too, but that's just a copy `find(...)` with extra steps...

**Comment by korydraughn** [comment]
Absolutely. Providing something like `find_copy` sounds like a good function too.

It would simply make a copy of the object and return it. Then the issue goes away.

I think having an extract function and find+copy function would improve the API.

**Comment by trel** [comment]
so each dereference decrements the counter?

```cpp
  // Use the object just like a pointer.
  p1->member_a = // value.
  assert(p2.use_count() == 1);
  (*p1).member_a;
  assert(p2.use_count() == 0);
```

**Comment by trel** [comment]
seems very elegant - will definitely need some quality words to describe how to hold it to give programmers confidence it is doing what they want/think.

**Comment by korydraughn** [comment]
Unless @MartinFlores751 finds something better, I think the proxy object approach is the winner.

**Comment by korydraughn** [comment]
> so each dereference decrements the counter?
> 
> ```c++
>   // Use the object just like a pointer.
>   p1->member_a = // value.
>   assert(p2.use_count() == 1);
>   (*p1).member_a;
>   assert(p2.use_count() == 0);
> ```

Just want to highlight that the code you posted is not the behavior I'm describing.

In the code you shared, the use count would be greater than zero for both assertions because `p1` exists. The internal count would decrement when the proxy object is destructed.

**Comment by korydraughn** [comment]
Bringing back into the 4.3.1 milestone.

**Comment by trel** [comment]
> In the code you shared, the use count would be greater than zero for both assertions because p1 exists.

but that... means your example... still isn't quite right?

```
process_stash::exists(key); // Returns true.
process_stash::erase(key);  // Erases the key-value pair because pN.use_count() == 0.
```

**Comment by korydraughn** [comment]
By the time you reach the lines doing the existence check, `p1` and `p2` will have already been destructed. The destruction of those objects causes the use count to decrement.

The reason they are destructed is because they are tied to the scope created by the curly braces.

**Comment by trel** [comment]
I guess I'm still confused about the first `exists()` returning `true`.

But happy to have the documentation have a good exhaustive example.

**Comment by MartinFlores751** [comment]
I think the solution @korydraughn has is probably the way to go.

Bonus Thoughts:
- One other case I thought up was where two threads hold the same data & both write to it. This can be solved by limiting access using the same counting in @korydraughn 's solution... We would need to decide if we want to block or not in that case, I believe.

- Another consideration is the `erase(...)` behavior... Should it forget like in the example given or "mark" it for deletion when it's unused?

Other than that, I think there's no other race conditions? Just need to be careful in implementation.

**Comment by korydraughn** [comment]
> * One other case I thought up was where two threads hold the same data & both write to it. This can be solved by limiting access using the same counting in @korydraughn 's solution... We would need to decide if we want to block or not in that case, I believe.

Hmm. Should that be the responsibility of the proces stash? That's what I'm wondering right now?

> * Another consideration is the `erase(...)` behavior... Should it forget like in the example given or "mark" it for deletion when it's unused?

What are the pros/cons/trade-offs?

**Comment by korydraughn** [comment]
After some discussion, we've decided to copy the process_stash implementation into the HTTP API repo and experiment with the solutions presented in this thread.

Once we determine which solution works best, we'll update the implementation that ships with the iRODS server.

Bumping to 4.3.2 milestone.

**Comment by trel** [comment]
Oh nice.   An elegant process for the process stash.

**Comment by korydraughn** [comment]
Here's yet another idea ...

We add an `update` function which takes a callable. The callable would take the object associated with the key and allow it to be modified in a safe way. This would play well with `find_copy`.

Here's an example:
```cpp
bool executed = irods::process_stash::update(key, [](boost::any& _object) {
    // If "key" is valid, the underlying data managed by process_stash is locked
    // and "_object" is passed to the lambda allowing safe writes to it. 
});
```

The issue with this idea is `find` still returns a pointer and it may be confusing to users. One way to fix that is to provide single-threaded API functions and multi-threaded API functions. Perhaps `process_stash` drops all thread-related logic and a new interface is introduced beside it that supports multi-threaded environments.

**Comment by korydraughn** [comment]
Another one ...

The idea here is locking the process stash is entirely up to the developer. By moving the responsibility of acquiring the lock out of the sibling operations, devs can carry out more operations while holding the lock. Like other solutions, this still requires the process stash interface to be changed if this approach seems reasonable.
```cpp
const auto lock_handle = irods::process_stash::lock(); // Can even have a try_lock.

// Use the process stash freely.
// When "lock_handle" is destructed, the write lock is released.
```

I'm starting to think the process stash shouldn't contain any locks at all. Then devs can build wrappers around it if they want to use it in a multi-threaded application.
